libtiledbsoma/src/reindexer/reindexer.cc

```diff
@@ -0,0 +1,159 @@
+/**
+ * @file   reindexer.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the a pybind11 api in the SOMA C++ library.
+ */
+
+#include "reindexer.h"
+#include <thread_pool/thread_pool.h>
+#include <unistd.h>
+#include "khash.h"
+#include "soma/enums.h"
+#include "soma/soma_array.h"
+#include "utils/arrow_adapter.h"
+#include "utils/common.h"
+#include "utils/logger.h"
+
+// Typedef for a 64-bit khash table
+KHASH_MAP_INIT_INT64(m64, int64_t)
+
+namespace tiledbsoma {
+
+void IntIndexer::map_locations(const int64_t* keys, int size, int threads) {
+    if (size == 0) {
+        return;
+    }
+    if (size < 10) {
+        threads = 1;
+    }
+    if (size < threads)
+        throw std::runtime_error(
+            "The number of keys " + std::to_string(size) +
+            " must be larger than the number of threads " +
+            std::to_string(threads) + " .");
+
+    LOG_DEBUG(fmt::format(
```

Comment:
We might want to consider adding `LOG_DEBUG` lines for codecov to ignore in `pyproject.toml`. I'm not sure if these debugging lines are enabled when running CI. If they aren't then we could probably get rid of a lot of false positives this way.

Reply:
Since codecov is complaining, it looks like we should add unit tests that directly test the C++ reindexer code.

Reply:
Yeah, I was actually thinking about that. I added that codecov support recently and it's biting me first. :)

Reply:
@nguyenv done